### ISSUE TITLE
Implement stratified resampling np.digitize in XLA

### ISSUE THREAD

**Issue by tel** [google/jax#2595]
Hi JAX team,

I'm interested in implementing efficient stratified resampling in JAX. This is a technique for variance reduction in sequential monte carlo algorithms (see section 2.2.1 in Naesseth, Lindsten, and Schön "Elements of Sequential Monte Carlo").

The algorithm provided looks like this

```python
# w.shape = (N, ) 
# x.shape = (N, )
# assert np.sum(w) = 1
def stratified_resampling(key, w, x) :
  N = w.shape[0]
  u = (np.arange(N) + jax.random.uniform(key, (N, )) / N
  bins = np.cumsum(w)
  return x[np.digitize(u, bins)]
```

although I'm also considering a version that doesn't actually do the shuffling in the last step (saving it for a "decoding" process later on which doesn't particularly need to be fast.

The immediate issue is that `np.digitize` is not implemented in JAX. As it were, I don't actually need the gradient for it, but as it's in the hot path and defeats the tracer it's causing some real pain. Even more ideally, I'd like to have something like the interface to `jax.random.categorical` where I can just use unnormalized logits.

How can I help implement at least tracing through `np.digitize`? Alternatively, is there a clever way of doing stratified resampling without using `np.digitize`?

Thanks

**Comment by terhorst** [comment]
@tel I had the same issue. `digitize` relies on `searchsorted`. The latter is not implemented in JAX either, but you can sort of hack it if you don't mind a few extra allocations and sorts, see https://github.com/google/jax/issues/2080#issuecomment-616234407.

**Comment by hawkinsp** [comment]
@jakevdp interested?
